docs/proposals/multi-tenant-networking.md

```diff
@@ -0,0 +1,194 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/multi-tenant-networking.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Multi-tenant networking
+
+## Abstract
+
+A proposal of multi-tenant networking by network providers.
+
+## Motivation
+
+The current kubernetes network design works great for connecting containers and building services, but it doesn’t work on many use-cases, such as container network isolation, ip address overlapping and so on. So this proposal adds network provider which adds multi-tenant networking for kubernetes.
+
+Goals of this design:
+
+* Add multi-tenant networking to support network isolation and ip address overlapping between containers
+* Support integration with SDN applications and external IPAMs by network provider
+* Works togather with current network model (if there is no network associated with the namespace, then pod networks of this namespace will work on current network model)
+
+## Use cases
+
+Multi-tenant networking is extremely useful for following use cases:
+
+* Network isotion between container networks
+* Manage kubernetes pod IPs by SDN applications
+* Integration with current IPAM tools
+
+## Proposed design
+
+To support multi-tenant networking, there are some changes to current kubernetes network design:
+
+* Instead of full communication between all containers, containers only on the same network are fullly communicated in this network model. Containers on different networks are not communicated directly, but they can communicate with each other via external endpoints.
+* Pods belongs to different network may overlap ip addresses. This is ok since networks are isolated.
+* Service ips are not virtual with network provider. They are allocated by network provider and load-balanced to specified pods.
+* Kube-proxy is network-aware, it only load-balances services whose namespace is with no network associated.
+* Each pod will receive a unique IP address from the network provider, and this IP will be configured for container by kubelet's network plugin, which works together with network controller.
+
+The main concept of multi-tenant networking is mapping network to namespaces, that is:
+* Each namespace is belong to one network, but network can be accross different namespaces.
```

Comment:
> @dcbw What good reason do you mean? And why it's not flexible?

Fundamentally it's the pod itself that gets added/removed from one or more networks.  The namespace is more about policies, constraints, and resource separation on a higher level.  I think the needs of networks and how pods get mapped to them are different than with namespaces.

For an example:

Namespace Production
    Pod Web: Network A
    Pod Database: Network A, Network Storage

Namespace Test
    Pod Web2: Network B
    Pod Database2: Network B, Network Storage

The current proposal has two problems as I see it (unless I badly misunderstand it!), which prevent the above configuration:

1. Namespaces can have one-and-only-one Network
2. Pods can only be attached to one network at a time, because Namespaces can have only one network, and because Pods can only exist in a single namespace

With labels (or some other mechanism to map Pods <-> Networks) we can support a more flexible mapping between the two, and ensure that eg Pod Web doesn't have access to the Storage network because there's no reason for it to have that access.

The drawback of labels is that untrusted Pod creators can specify them, so obviously we need some kind of access control.  That could be accomplished in a couple ways; two that come to mind are:

* Add an AllowedNetworks list to the Namespace object which gates which networks any pod started in that Namespace is allowed to use
* Or, use a "push-not-pull" model where the Network itself indicates which Namespaces it wishes to offer itself to, putting the access control on the Network object instead. I'm personally in favor of this one.

> If you map pods with networks, then network provider makes no sense. You can just do that by kubelet network plugin.

I think network providers still make sense; the proposal says "Network controller listens to creation/deletion/update events for networks, and tell network provider to configure them." and I don't think that has changed.  *Something* still needs to allocate network-global resources on each node when the network is found via etcd, and tear down those resources when the network disappears.  That would be a network provider, right?

Eg, say each network on each node gets a Linux bridge with the subnet specified in the NetworkSpec but with the ".1" address.  Kubernetes doesn't care how that gets configured because that's the job of the network provider.  The provider may need to set this bridge up immediately when the network is found, or it may need to set the bridge up when the first pod using that network is started on the node.

Likewise, something has to handle Services on a per-node basis, since these are somewhat independent of when pods start and stop.  That thing would most likely be the network provider.

> And it's very hard to meet network isolation use-cases by mapping networks via labels.

Network isolation simply means that two pods on different networks cannot talk to each other.  That's clearly still possible with labels if two pods give two different labels.  Obviously there needs to be more than that, since labels are specified by the Pod creator and untrusted.

But perhaps each Namespace gets a separate network by default that is isolated from all other networks, but pods can opt-in to other networks via labels or something like that.